### ISSUE TITLE
Is order preservation of `Enumerable.GroupBy` guaranteed?

### ISSUE THREAD

**Issue by colejohnson66** [dotnet/runtime#93857]
A few times, I've relied on `GroupBy` having the resulting groups preserve the ordering of the input elements. For example ([SharpLab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbj0JMwDoAZASygEcrdqBmE9WgdjwBvPETG90o8SNzi5RZgGdgAHgCyATwBKAUwDGAezBwAfETD6jcRUQC8RKDoDuACgCUU+URlevj1+gwRJhuMJ6+Yv4ugUTooeERUVxBXPGyEeJJQWRpGZnO0UEArLl5DgWYQQBspV4AvhwRCeKkXOralsYA2gC6ZjpQEAC2OmAAhgBGADY6duad1s1edADiYAYQAA4AQhouqKgWhsZ2ZgdHVnQAgh7peXQAyjozesAuAObrW6dEnxubdAAKgYrmBxns3G5Gr4AGZGHRjPQACyILk0umOcF6vy+myILCIA2Go0mM1uER8ZVIAE4XAASABEa3+IG8fy2XQIPWudQZUKW8jhFkRKLRHUxJEOC3xUBx/3JZTkNP2UsxdG2/Lu8jq4R1WvCqB4FxO6IWLhYwCIVyCFqIGo4dSAA)):

```csharp
using System;
using System.Collections.Generic;
using System.Linq;

public class Program
{
    public static void Main()
    {
        List<MyRecord> records = new()
        {
            new(2, 1),
            new(2, 2),
            new(3, 3),
            new(3, 4),
            new(2, 5),
            new(1, 6),
        };
        
        IEnumerable<MyRecord[]> enumerable = records
            .GroupBy(@record => @record.A)
            .Select(group => group.ToArray());
        foreach (MyRecord[] group in enumerable)
        {
            Console.WriteLine($"Group: {group[0].A}");
            foreach (MyRecord @record in group)
                Console.WriteLine(@record.B);
        }
    }
    
    public record MyRecord(int A, int B);
}
```

Output:
```
Group: 2
1
2
5
Group: 3
3
4
Group: 1
6
```

Such behavior is [not documented](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=net-7.0). Instead, we just are told that this group of methods "[g]roups the elements of a sequence" with no regard for ordering.

Taking a peek at the runtime's source code, we see that ordering is indeed preserved. `Enumerable.GroupBy` returns a `GroupedEnumerable<,>` which uses `Lookup<,>.Create`. `Create` iterates through the source enumerable, adding them one at a time, and the method it uses appears to preserve ordering (given a cursory glance).

Basically, is this property guaranteed by the runtime, or are you reserving the right to change this behavior in the future?

---

Tangent: are the `IGrouping` objects (returned here, not necessarily by an external API) guaranteed not to be empty sequences?

**Comment by eiriktsarpalis** [comment]
> Such behavior is [not documented](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=net-7.0). Instead, we just are told that this group of methods "[g]roups the elements of a sequence" with no regard for ordering.

While this is true for the case of LINQ-to-objects, for the case of IQueryable it would largely depend on the implementation of the underlying query provider. While I don't have specific evidence pointing to this, I wouldn't be surprised if certain NoSQL query providers provide nondeterministic ordering under certain circumstances (e.g. querying across partitions). So even though the LINQ-to-objects is stable w.r.t. ordering and this is likely never going to change, I don't think it would be wise to provide explicit guarantees about this.

> Tangent: are the IGrouping objects (returned here, not necessarily by an external API) guaranteed not to be empty sequences?

Yes.

**Comment by Clockwork-Muse** [comment]
> but I wouldn't be surprised if certain NoSQL query providers provide nondeterministic ordering under certain circumstances (e.g. when querying across partitions). 

It definitely doesn't for SQL query providers (where the underlying engines are notorious for giving apparently non-deterministic results when you least expect it without an `ORDER BY`).

**Comment by eiriktsarpalis** [comment]
Closing as answered.
